integrations/adobe-analytics/lib/index.js

```diff
@@ -676,13 +673,39 @@ function extractProperties(props, options) {
     options.contextValues
   ];
 
+  var topLevelProperties = ['messageId', 'anonymousId'];
+
+  var props = facade.properties();
+  if (propType === 'mergedPropContext') {
+    props = merge(trample(facade.properties()), trample(facade.context()));
+  } else if (propType === 'context') {
+    props = trample(facade.context());
+  }
+
   // Iterate through each variable mappings to find matching props
   for (var x = 0; x < mappings.length; x++) {
     each(match, mappings[x]);
   }
 
   function match(mappedValue, mappedKey) {
     var value = dot(props, mappedKey);
+    // On track and page calls the propType here will be empty. On video HB calls the propTyp will
+    // be 'mergedPropContext'. On those events we use  the extracted properties from this fxn to explicitly
+    // map link vars on `window.s` object.
+    if (topLevelProperties.includes(mappedKey) && propType !== 'context') {
+      value = facade.proxy(mappedKey);
+    }
+
+    // When we are checking the context object for segment property mappings (updateContextData)
+    // we only want set top level fields to Context Data Variables.
+    var contextValueKeys = Object.keys(options.contextValues);
```

Comment:
So `extractProperties` is used many ways. 
1. For `track()` and `page()` calls we pass this fxn the `properties` object and use the result to update link vars on the `window.s` object
2. For HB events we access both the `properties` and `context` object and use the result to update custom content metadata on video events. 
3. When updating context data. We update context data on `track()`, `page()`, and all e-commerce events. These `contetxtValues` from our destination settings  are `segmentContextField: adobeContextDataField`. For these we have always looked in just the context object. There is a bug in our implementation that unfortunately now must be maintained [that you can see here](https://github.com/segmentio/analytics.js-integrations/pull/465/files/2c34180bc58909e687d4566ec2cc4d9ab00c6526#diff-e7435b76f43719e786ff94a43acc1f6eR569-L574). When we go to update context data fields we actually also look at eVars, props, hVars, lVars etc. These usually always return undefined because their `mappedKey` is defined in the properties object not the context object. 

Basically if the toplevelProperty is `anonymousId` or `messageId` and we are NOT setting context data values we can set `value = facade.proxy(mappedKey);`. Otherwise we only want to set that if the a customer defines they want that top-level mapping to update Context Data Fields by checking their `contextValues` settings.  Otherwise we could incorrectly set `contextData.eVar5` when we update context data.

This code definitely needs a larger refactor IMO, but this will solve the immediate need from Fox and Intuit.

Reply:
Got it. Thanks for the in-depth write-up. I totally overlooked the bug we need to maintain. 

> This code definitely needs a larger refactor IMO, but this will solve the immediate need from Fox and Intuit.

Agreed. I think this is all 🟢 to merge.